stdlib/public/core/StringUTF16View.swift

```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
   }
 }
 
+// FIXME: This is for debugging only; remove before merging.
+extension String.Index: CustomStringConvertible {
+  @_alwaysEmitIntoClient
+  @inline(never)
+  public var description: String {
+    var d = "Index("
+      d += "offset: \(_encodedOffset)"
+    if transcodedOffset != 0 {
+      d += "+\(transcodedOffset)"
+    }
+    if let stride = characterStride {
+      d += ", stride: \(stride)"
+    }
+    #if false // These aren't a thing yet
+    if _isCharacterAligned {
+      d += ", character"
+    } else if _isScalarAligned {
+      d += ", scalar"
+    }
+    if _rawBits & 0x4 != 0 {
+      d += ", utf8"
+    }
+    if _rawBits & 0x8 != 0 {
+      d += ", utf16"
+    }
+    #endif
+    d += ")"
+    return d
+  }
+}
+
 extension String.UTF16View {
+  
+  @inline(__always)
+  internal func _utf16Length<U: SIMD, S: SIMD>(
+    readPtr: inout UnsafeRawPointer,
+    endPtr: UnsafeRawPointer,
+    unsignedSIMDType: U.Type,
+    signedSIMDType: S.Type
+  ) -> Int where U.Scalar == UInt8, S.Scalar == Int8 {
+    var utf16Count = 0
+    
+    while readPtr + MemoryLayout<U>.stride < endPtr {
+      //Find the number of continuations (0b10xxxxxx)
+      let sValue = readPtr.load(as: S.self)
+      let continuations = S.zero.replacing(with: S.one, where: sValue .< -65 + 1)
+      let continuationCount = Int(continuations.wrappedSum())
+            
+      //Find the number of 4 byte code points (0b11110xxx)
+      let uValue = readPtr.load(as: U.self)
+      let fourBytes = U.zero.replacing(with: U.one, where: uValue .>= 0b11110000)
+      let fourByteCount = Int(fourBytes.wrappedSum())
+            
+      utf16Count &+= (U.scalarCount - continuationCount) + fourByteCount
+      
+      readPtr += MemoryLayout<U>.stride
+    }
+    
+    return utf16Count
+  }
+  
+  @inline(__always)
+  internal func _utf16Distance(from start: Index, to end: Index) -> Int {
+    _internalInvariant(end.transcodedOffset == 0 || end.transcodedOffset == 1)
+    return (end.transcodedOffset - start.transcodedOffset) + _guts.withFastUTF8(
```

Comment:
This handling of transcoded offsets makes some sense to me and passes all tests, but I'd appreciate eyes on it, since it's the bit I'm least confident in